<!DOCTYPE html>
<html>
    <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <meta name="author" content="Joseph D. Romano" />
    
    
    <title>Lab Publications — Romano Lab @ Penn</title>

    <!--Icons-->
    <script src="https://use.fontawesome.com/04f6e4e6fe.js"></script>
    
<link rel="icon" href="/favicon.ico" sizes="any">
<link rel="icon" href="/favicon.svg" type="image/svg+xml">
<link rel="apple-touch-icon" href="/apple-touch-icon.png">
<link rel="manifest" href="/manifest.webmanifest">
  

    <!--p5js CDN-->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/1.6.0/p5.min.js"></script>

    <link rel="stylesheet" href="/_includes/assets/css/style.css"/>
    
    
    <!--Semantic metadata-->
    <script type="application/ld+json">
    {
        "@context": "http://schema.org/",
        "@type": "Organization",
        "name": "The Romano Lab at the University of Pennsylvania",
        "url": "https://romanolab.org/",
        "address": {
            "@type": "PostalAddress",
            "addressLocality": "Philadelphia, Pennsylvania, USA",
            "postalCode": "09104",
            "streetAddress": "403 Blockley Hall 423 Guardian Drive"
        },
        "email": "joseph.romano@pennmedicine.upenn.edu",
        "telephone": "+1-802-558-5839",
        "member": [
            "https://jdr.bio"
        ],
        "sameAs": [
            "https://www.wikidata.org/wiki/Q116036288"
        ]
    }
    </script>
    <meta property="og:title" content="Lab Publications — Romano Lab @ Penn" />
    <meta property="og:type" content="website" />
    <meta property="og:url" content="/publications.html" />
    <meta property="og:site_name" content="The Romano Lab at the University of Pennsylvania" />
    <meta property="og:locale" content="en_US" />
    <meta name="robots" content="index, follow" />

</head>
    
    <body>
        <div class="container">
            <header>
    <nav>
        <ul class="nav-list">
            <li class="nav-item" id="logo"><a href="/index.html" >Romano Lab</a> @ <a href="https://upenn.edu" target="_blank" id="pennlink">Penn</a></li>
            <li class="nav-item"><a href="/research.html" >Research</a></li>
            <li class="nav-item"><a href="/people.html" >People</a></li>
            <li class="nav-item"><a href="/publications.html" aria-current="page">Publications</a></li>
            <li class="nav-item"><a href="/contact.html" >Contact</a></li>
            
        </ul>
    </nav>
</header>
            
            
            
            <main>
                <div class="sidebar"></div>
                <div id="content">
                    

<h2>Lab Publications (newest first)</h2>
<p>
    The following publications were either written by members of the lab, or were contributed to significantly by a lab member as a coauthor.
</p>
<div class="pub-entry">
<p class="pub-title"><b>The Alzheimer’s Knowledge Base – A knowledge graph for therapeutic discovery in Alzheimer’s Disease research</b></p>

<p class="pub-authors">Romano JD, Truong V, Kumar R, Venkatesan M, Graham BE, Hao Y, Matsumoto N, Li X, Wang Z, Ritchie M, Shen L, and Moore JH.</p>


<p class="pub-journal"><i>Journal of Medical Internet Research. (in press).</i>. </p>

<div class="pub-buttons">
    
    
    
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Discovering Venom-Derived Drug Candidates using Differential Gene Expression</b></p>

<p class="pub-authors">Romano JD, Li H, Napolitano T, Realubit R, Karan C, Holford M, and Tatonetti NP.</p>


<p class="pub-journal"><i>Toxins. 2023;15(7):451.</i>. (DOI: <a href="https://doi.org/10.3390/toxins15070451" target="_blank">10.3390/toxins15070451</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.mdpi.com/2072-6651/15/7/451';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/37505720';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10467105';" target="_blank">PubMed Central</button>
    
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Exploring genetic influences on adverse outcome pathways using heuristic simulation and graph data science</b></p>

<p class="pub-authors">Romano JD, Mei L, Senn J, Moore JH, and Mortensen HM.</p>


<p class="pub-journal"><i>Computational Toxicology. 2023;25:100261</i>. (DOI: <a href="https://doi.org/10.1016/j.comtox.2023.100261" target="_blank">10.1016/j.comtox.2023.100261</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.sciencedirect.com/science/article/abs/pii/S2468111323000026';" target="_blank">Journal link</button>
    
    
    
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Omics Methods in Toxins Research - A toolkit to Drive the Future of Scientific Inquiry</b></p>

<p class="pub-authors">Romano JD.</p>


<p class="pub-journal"><i>Toxins. 2022;14(11):761</i>. (DOI: <a href="https://doi.org/10.3390/toxins14110761" target="_blank">10.3390/toxins14110761</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.mdpi.com/2072-6651/14/11/761';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/36356011';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9692551';" target="_blank">PubMed Central</button>
    
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Knowledge-guided deep learning models of drug toxicity improve interpretation</b></p>

<p class="pub-authors">Hao Y, Romano JD, and Moore JH.</p>


<p class="pub-journal"><i>Patterns. 2022 Aug 24;3(9):100565</i>. (DOI: <a href="https://doi.org/10.1016/j.patter.2022.100565" target="_blank">10.1016/j.patter.2022.100565</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.cell.com/patterns/fulltext/S2666-3899(22)00176-3';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/36124309';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9481960';" target="_blank">PubMed Central</button>
    
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Automating predictive toxicology using ComptoxAI</b></p>

<p class="pub-authors">Romano JD, Hao Y, Moore JH, and Penning TM.</p>


<p class="pub-journal"><i>Chem. Res. Toxicol. 2022;35(8):1370-1382</i>. (DOI: <a href="https://doi.org/10.1021/acs.chemrestox.2c00074" target="_blank">10.1021/acs.chemrestox.2c00074</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://pubs.acs.org/doi/10.1021/acs.chemrestox.2c00074';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/35819939';" target="_blank">PubMed</button>
    
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;cluster=15039439195538845252&amp;hl=en';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>PMLB v1.0: An open source dataset collection for benchmarking machine learning methods</b></p>

<p class="pub-authors">Romano JD, Le TT, La Cava W, Gregg JT, Goldberg DJ, Chakraborty P, Ray NL, Himmelstein D, Fu W, and Moore JH.</p>


<p class="pub-journal"><i>Bioinformatics. 2022 Feb 1;38(3):878-880</i>. (DOI: <a href="https://doi.org/10.1093/bioinformatics/btab727" target="_blank">10.1093/bioinformatics/btab727</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://academic.oup.com/bioinformatics/article/38/3/878/6408434';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/34677586';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8756190';" target="_blank">PubMed Central</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;cluster=17133623133497859665&amp;hl=en';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Improving QSAR Modeling for Predictive Toxicology using Publicly Aggregated Semantic Graph Data and Graph Neural Networks</b></p>

<p class="pub-authors">Romano JD, Hao Y, and Moore JH.</p>


<p class="pub-journal"><i>Pac Symp Biocomput. 2022;27:187-198</i>. (DOI: <a href="https://doi.org/10.1142/9789811250477_0018" target="_blank">10.1142/9789811250477_0018</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.worldscientific.com/doi/abs/10.1142/9789811250477_0018';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/34890148';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8714189';" target="_blank">PubMed Central</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;cluster=12882306140115916832&amp;btnI=1&amp;hl=en';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>The promise of automated machine learning for the genetic analysis of complex traits</b></p>

<p class="pub-authors">Manduchi E, Romano JD, and Moore JH.</p>


<p class="pub-journal"><i>Human Genetics. 2022;141:1529-1544</i>. (DOI: <a href="https://doi.org/10.1007/s00439-021-02393-x" target="_blank">10.1007/s00439-021-02393-x</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://link.springer.com/article/10.1007/s00439-021-02393-x';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/34713318';" target="_blank">PubMed</button>
    
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;cluster=5716636089984178836&amp;btnI=1&amp;hl=en';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>TPOT-NN: Augmenting tree-based automated machine learning with neural network estimators</b></p>

<p class="pub-authors">Romano JD, Le TT, Fu W, and Moore JH.</p>


<p class="pub-journal"><i>Genetic Programming and Evolvable Machines. 2021;22:207-227</i>. (DOI: <a href="https://doi.org/10.1007/s10710-021-09401-z" target="_blank">10.1007/s10710-021-09401-z</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://link.springer.com/article/10.1007/s10710-021-09401-z';" target="_blank">Journal link</button>
    
    
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;cluster=1062939493121597080&amp;btnI=1&amp;hl=en';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Ten Simple Rules for Writing a Paper About Scientific Software</b></p>

<p class="pub-authors">Romano JD and Moore JH.</p>


<p class="pub-journal"><i>PLoS Comput Biol. 2020 Nov 12;16(11):e1008390</i>. (DOI: <a href="https://doi.org/10.1371/journal.pcbi.1008390" target="_blank">10.1371/journal.pcbi.1008390</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008390';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/33180774';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7660560';" target="_blank">PubMed Central</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;cluster=14945349318462945153&amp;btnI=1&amp;hl=en';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Embedding covariate adjustments in tree-based automated machine learning for biomedical big data analysis</b></p>

<p class="pub-authors">Manduchi E, Fu W, Romano JD, Ruberto S, and Moore JH.</p>


<p class="pub-journal"><i>BMC Bioinformatics. 2020;21:430</i>. (DOI: <a href="https://doi.org/10.1186/s12859-020-03755-4" target="_blank">10.1186/s12859-020-03755-4</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-020-03755-4';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/32998684';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7528347';" target="_blank">PubMed Central</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;cluster=11711136781459864682&amp;btnI=1&amp;hl=en';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Informatics and Computational Methods in Natural Product Drug Discovery: A Review and Perspectives</b></p>

<p class="pub-authors">Romano JD and Tatonetti NP.</p>


<p class="pub-journal"><i>Front. Genet., 30 April 2019</i>. (DOI: <a href="https://doi.org/10.3389/fgene.2019.00368" target="_blank">10.3389/fgene.2019.00368</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.frontiersin.org/articles/10.3389/fgene.2019.00368/full';" target="_blank">Journal link</button>
    
    
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?cluster=11461611408044634283&amp;hl=en&amp;as_sdt=0,33&amp;as_vis=1';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>A Decade of Translational Bioinformatics: A Retrospective Analysis of &quot;Year-in-Review&quot; Presentations</b></p>

<p class="pub-authors">Romano JD, Bernauer M, McGrath S, Nagar SD, and Freimuth R.</p>


<p class="pub-journal"><i>AMIA 2019 Informatics Summit</i>. </p>

<div class="pub-buttons">
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/27570672';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5001765';" target="_blank">PubMed Central</button>
    
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Using a Novel Ontology to Inform the Discovery of Therapeutic Peptides from Animal Venoms</b></p>

<p class="pub-authors">Romano JD and Tatonetti NP.</p>


<p class="pub-journal"><i>AMIA Summits on Translational Science Proceedings, 2019</i>. </p>

<div class="pub-buttons">
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/27570672';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5001765';" target="_blank">PubMed Central</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;hl=en&amp;cluster=13812255261821036462';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Systems Biology Approaches for Identifying Adverse Drug Reactions and Elucidating their Underlying Biological Mechanisms</b></p>

<p class="pub-authors">Boland MR, Jacunski A, Lorberbaum T, Romano JD, Moskovitch R, and Tatonetti NP.</p>


<p class="pub-journal"><i>WIREs Systems Biology and Medicine, 8(2), 104-122 (2016)</i>. (DOI: <a href="https://doi.org/10.1002/wsbm.1323" target="_blank">10.1002/wsbm.1323</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://onlinelibrary.wiley.com/doi/full/10.1002/wsbm.1323';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/26559926';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4760887';" target="_blank">PubMed Central</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;hl=en&amp;cluster=14174248577525979029';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>VenomKB, A new knowledge base for facilitating the validation of putative venom therapies</b></p>

<p class="pub-authors">Romano JD and Tatonetti NP.</p>


<p class="pub-journal"><i>Scientific Data, 2, 150065 (2015)</i>. (DOI: <a href="https://doi.org/10.1038/sdata.2015.65" target="_blank">10.1038/sdata.2015.65</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.nature.com/articles/sdata201565';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/26601758';" target="_blank">PubMed</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4658572';" target="_blank">PubMed Central</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;hl=en&amp;cluster=10749764197677184379';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

<div class="pub-entry">
<p class="pub-title"><b>Adapting Simultaneous Analysis Phylogenomic Techniques to Study Complex Disease Gene Relationships</b></p>

<p class="pub-authors">Romano JD, Tharp WG, and Sarkar IN.</p>


<p class="pub-journal"><i>Journal of Biomedical Informatics, 54, 10-38 (2014)</i>. (DOI: <a href="https://doi.org/10.1016/j.jbi.2015.01.002" target="_blank">10.1016/j.jbi.2015.01.002</a>)</p>

<div class="pub-buttons">
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.sciencedirect.com/science/article/pii/S1532046415000040';" target="_blank">Journal link</button>
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://www.ncbi.nlm.nih.gov/pubmed/25592479';" target="_blank">PubMed</button>
    
    
    
    <button class="pub-button-journal" onclick="window.location.href = 'https://scholar.google.com/scholar?oi=bibs&amp;hl=en&amp;cluster=2933870931495223365';" target="_blank">Google Scholar</button>
    
    
</div>
</div>

                </div>
                <footer>
    <div class="footer-text">
        <p><i>Copyright &copy; 2023 by <a href="http://jdr.bio" id="jdrlink">Joseph D. Romano</a> and the Romano Lab</i></p>
    </div>
</footer>
            </main>
            
            
        </div>
        
        <script>
            /**
 * graph.js
 * Copyright (c) 2022 by Joseph D. Romano
 * 
 * Simple routines for drawing random graphs
 */


// Global variables

const globals = {
    'nodeRadius': 4,
    'nodeSpacing': 25,
    'edgeWidth': 2,
    'edgeColor': '#808080',
    'breathingRoom': 8
};

const canvas = document.querySelector('canvas');
const ctx = canvas.getContext('2d');

canvas.width = 800;
canvas.height = 250;

const colors = [
    '#e91e63',
    '#fa750f',
    '#ffeb3b',
    '#4caf50',
    '#02b3e4',
    '#c341d8'
]; 

/**
 * Retrieve a random sample of an array by performing the Fisher-Yates shuffle
 * and then taking a slice of the desired size.
 * 
 * See: https://stackoverflow.com/a/11935263/1730417
 * 
 * @param {Array} arr An array from which to draw samples
 * @param {Integer} n Number of samples to draw from `arr`
 */
function getRandomSample(arr, n) {
    let shuffled = arr.slice(0);
    let i = arr.length;
    let temp, idx;

    while (i--) {
        idx = Math.floor((i+1) * Math.random());
        temp = shuffled[idx];
        shuffled[idx] = shuffled[i];
        shuffled[i] = temp;
    }
    
    return shuffled.slice(0, n);
}


// Data structures

class Node {
    constructor(x, y) {
        this.x = x;
        this.y = y;

        this.left   = this.x - globals.nodeSpacing / 2 + globals.nodeRadius;
        this.top    = this.y - globals.nodeSpacing / 2 + globals.nodeRadius;
        this.right  = this.x + globals.nodeSpacing / 2 - globals.nodeRadius;
        this.bottom = this.y + globals.nodeSpacing / 2 - globals.nodeRadius;

        this.abs_x = this.left + globals.breathingRoom + (this.right - this.left) * Math.random();
        this.abs_y = this.top + globals.breathingRoom + (this.bottom - this.top) * Math.random();

        this.color = colors[Math.floor(Math.random() * colors.length)];
    }

    drawNode() {
        ctx.fillStyle = this.color;
        ctx.beginPath();
        ctx.arc(this.abs_x, this.abs_y, globals.nodeRadius, 0, 2 * Math.PI);
        ctx.fill();
    }

}


class Edge {
    constructor(startNode, endNode) {
        this.startNode = startNode;
        this.endNode = endNode;

        this.x1 = this.startNode.abs_x;
        this.x2 = this.endNode.abs_x;

        this.y1 = this.startNode.abs_y;
        this.y2 = this.endNode.abs_y;
    }

    drawEdge() {
        ctx.globalAlpha = 1.0;
        ctx.strokeStyle = globals.edgeColor;
        ctx.edgeWidth = globals.edgeWidth;
        ctx.beginPath();
        ctx.moveTo(this.x1, this.y1);
        ctx.lineTo(this.x2, this.y2);
        ctx.stroke();
    }

    getLength() {
        return Math.sqrt(
            Math.pow((this.x1 - this.x2), 2) + Math.pow((this.y1 - this.y2), 2)
        );
    }
}

class Graph {
    /**
     * 
     * @param {canvas} cnv <canvas> object where the graph will be drawn
     */
    constructor(cnv) {
        this.nodes = Array();
        this.edges = Array();
        this.orphanNodes = Array();

        this.canvas = cnv
        this.canvas_height = cnv.height;
        this.canvas_width = cnv.width;

        this.usable_height = this.canvas_height - (globals.breathingRoom * 2);
        this.usable_width = this.canvas_width - (globals.breathingRoom * 2)
        
        //this.max_path_len = Math.sqrt(Math.pow(this.canvas_height, 2) + Math.pow(this.canvas_width, 2));
        this.max_path_len = Math.sqrt(Math.pow(this.canvas_height, 2) + Math.pow(this.canvas_width, 2)) / 5.;
    }

    /**
     * 
     * @param {Integer} canvas_x x-coordinate within the associated canvas
     * @param {Integer} canvas_y y-coordinate within the associated canvas
     */
    addNode(canvas_x, canvas_y) {
        let thisNode = new Node(canvas_x, canvas_y)
        this.nodes.push(thisNode)
        this.orphanNodes.push(thisNode);
    }

    addRandomNodes(n_nodes) {
        let random_node_coords = Array();
        
        for(var i=0; i < n_nodes; i++) {
            random_node_coords.push([
                Math.random() * this.usable_width,
                Math.random() * this.usable_height
            ]);
        }

        random_node_coords.forEach(nc => {
            this.addNode(nc[0], nc[1])
        });
    }

    addEdge(startNodeIndex, endNodeIndex) {
        const startNode = this.nodes[startNodeIndex];
        const endNode = this.nodes[endNodeIndex];
        
        this.edges.push(new Edge(startNode, endNode));
        if (this.orphanNodes.includes(startNode)) {
            this.orphanNodes.splice(this.orphanNodes.indexOf(startNode), 1);
        }
        if (this.orphanNodes.includes(endNode)) {
            this.orphanNodes.splice(this.orphanNodes.indexOf(endNode), 1);
        }
    }

    addRandomEdges(n_edges) {
        let randomSample;
        let acceptProb;

        // in its most basic version, we simply randomly select pairs of nodes
        // for(var i=0; i < n_edges; i++) {
        //     randomSample = getRandomSample(this.nodes, 2);
        //     this.edges.push(new Edge(randomSample[0], randomSample[1]));
        // }

        // here, we accept random samples with probability inversely 
        // proportional to the length of the resulting edge
        // for(var i=0; i < n_edges; i++) {
        //     randomSample = getRandomSample(this.nodes, 2);
        //     let candidateEdge = new Edge(randomSample[0], randomSample[1]);
        //     acceptProb = 1 - (candidateEdge.getLength() / this.max_path_len);
        //     if (Math.random() <= acceptProb) {
        //         this.edges.push(candidateEdge);
        //     } else {
        //         console.log("REJECT with probability ", acceptProb);
        //         --i;
        //     }
        // }

        // Same as above, but this time we adjust probability based on logistic
        // curve to more strongly encourage short edges
        // for(var i=0; i < n_edges; i++) {
        //     randomSample = getRandomSample(this.nodes, 2);
        //     let candidateEdge = new Edge(randomSample[0], randomSample[1]);
        //     // acceptProb = 1 - (candidateEdge.getLength() / this.max_path_len);
        //     acceptProb = 1 / (Math.pow(Math.E, 2*(candidateEdge.getLength() / this.max_path_len))+1);
        //     if (Math.random() <= acceptProb) {
        //         this.edges.push(candidateEdge);
        //     } else {
        //         console.log("REJECT with probability ", acceptProb);
        //         --i;
        //     }
        // }

        // Still not quite strong enough.
        for(var i=0; i < n_edges; i++) {
            randomSample = getRandomSample(this.nodes, 2);
            let candidateEdge = new Edge(randomSample[0], randomSample[1]);
            //acceptProb = 1 / (Math.pow(Math.E, 2*(candidateEdge.getLength() / this.max_path_len))+1);
            acceptProb = 1-(Math.tanh(candidateEdge.getLength() / this.max_path_len))
            if (Math.random() <= acceptProb) {
                this.edges.push(candidateEdge);
            } else {
                --i;
            }
        }
    }

    drawGraph() {
        // We draw the edges first, so their tips sit 'behind' the nodes
        this.edges.forEach(e => e.drawEdge());
        this.nodes.forEach(n => n.drawNode());
    }
}


// Draw a random graph
let G = new Graph(canvas);

function start() {
    // drawGraph(node_coords, adjacencies);
    G.addRandomNodes(120);
    G.addRandomEdges(200);

    // get rid of the orphan nodes
    // G.orphanNodes.forEach((o_n, i) => {
    //     G.nodes.splice(G.nodes.indexOf(o_n), 1);
    // })

    // ctx.scale(0.5, 0.5);
    G.drawGraph();
    // ctx.scale(2, 2);
    // G.drawGraph();
}

window.setTimeout(start, 100);
        </script>
    </body>
</html>